gp2-stress-experiment.ipynb

```diff
@@ -0,0 +1,933 @@
+   "source": [
+    "def stitch_data(variable_to_stitch):\n",
+    "    print(f\"Stitching: {variable_to_stitch}\")\n",
+    "    stitched[variable_to_stitch] = imaging.stitch(ds[variable_to_stitch],\n",
+    "                                                                           frame_parameters=frame_parameters,\n",
+    "                                                                           frame_shifts=frame_shifts,\n",
+    "                                                                           rebin_parameters=rebin_parameters)\n",
+    "\n",
+    "for var_name in ds.keys():\n",
+    "    stitch_data(var_name)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "tof = stitched.coords['tof']\n",
+    "bin_widths = tof['tof', 1:] - tof['tof', :-1]\n",
+    "\n",
+    "integrated = sc.Dataset()\n",
+    "for dim in stitched.coords.keys():\n",
+    "    integrated.coords[dim] = stitched.coords[dim]\n",
+    "\n",
+    "for k in stitched.keys():\n",
+    "    integrated[k] = sc.sum(stitched[k], \"tof\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "\n",
+    "num_spectra = stitched.coords[\"spectrum\"].shape[0]\n",
+    "bank_width = int(np.sqrt(num_spectra))\n",
+    "integrated = operations.integrate(stitched)\n",
+    "integrated_neighbour_spectra = operations.generate_neighbouring_spectra(num_spectra=num_spectra,\n",
+    "                                                                        bank_width=bank_width)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def mask_non_sample_regions(integrated_dataset, var_name):\n",
+    "    integrated_spectra = integrated_dataset[var_name]\n",
+    "    spectra_masks = [False] * num_spectra\n",
+    "    for i, spec_val in enumerate(integrated_spectra.values):\n",
+    "        if spec_val > masking_threshold:\n",
+    "            spectra_masks[i] =  True\n",
+    "\n",
+    "    for i, adj_spectra in integrated_neighbour_spectra.items():\n",
+    "        if all(spectra_masks[i] == True for i in adj_spectra):\n",
+    "            spectra_masks[i] = True  # Force to true as neighbours true\n",
+    "\n",
+    "        if all(spectra_masks[i] == False for i in adj_spectra):\n",
+    "            spectra_masks[i] = False\n",
+    "\n",
+    "    stitched.masks[var_name] = sc.Variable([\"spectrum\"], values=spectra_masks)\n",
+    "\n",
+    "for k in integrated.keys():\n",
+    "    if k == \"reference\":\n",
+    "        continue\n",
+    "\n",
+    "    print(f\"Masking non-sample regions in {k}\")\n",
+    "    integrated[k] = integrated[k] / integrated[\"reference\"]\n",
```

Comment:
Shame there is no Python warning we can enable for this (since it's undetectable in the binding layer), as this is an easy mistake to make and has large performance implications

Reply:
I made this last week: https://scipp.github.io/user-guide/tips-tricks-and-anti-patterns.html.